### DISCUSSION TITLE
Text Trimming and Tool Tips

### DISCUSSION THREAD

**Discussion by nixtar** [AvaloniaUI/Avalonia#14521] [Q&A] [answered]
Hey,

I have a situation where I want to use Text Trimming on a Text Block that displays a users name in the event they have a long name.

But I want to have a tooltip with the users full name that they can mouse over to see.
However, I don't want the tooltip to be set if the Text Block is not being trimmed.

I was thinking a style that sets the tooltip only when the trimming is active, however I don't see an obvious way to detect if trimming is happening.

Anyone got any ideas?

Thanks,
Nick

**Comment by maxkatz6** [comment, marked as answer]
Check like this should work:
```
var isTrimmed = textBlock.TextTrimming != TextTrimming.None && textBlock.TextLayout.TextLines.Any(x => x.HasCollapsed);
```

**Reply by nixtar** [reply]
Ahh TextLayout is what I was missing.
I was trying to come up with a solution in the xaml/styling only.

I've put the following in the code behind for the window that contains the Text Block:
```c#
private void TextBlock_OnLoaded(object? sender, RoutedEventArgs e)
{
    if (sender is not TextBlock textBlock) return;
    var isTrimmed = textBlock.TextTrimming != TextTrimming.None && textBlock.TextLayout.TextLines.Any(x => x.HasCollapsed);

    if (isTrimmed)
    {
        // Add method handles the case where the class is already added
        textBlock.Classes.Add("Trimmed");
    }
    else
    {
        textBlock.Classes.Remove("Trimmed");
    }
}

```

Then in the windows styles:
```xaml
<Style Selector="TextBlock.Trimmed">
  <Setter Property="ToolTip.Tip" Value="{Binding RelativeSource={RelativeSource Self}, Path=Text}"/>
</Style>
```

Do you know if it's possible to use a fancy style selector for this? I assume the `TextLines` linq part might be an issue?

**Reply by maxkatz6** [reply]
No, I don't think there would be a selector for this.

Also, your "Trimmed" class is only evaluated when control is loaded. But trimming state might be changed on SizeChanged or TextChanged as well.

**Reply by nixtar** [reply]
Yep I thought about this however, in my exact use case the view model/view does not change for the lifetime of the view/window.

It doesn't appear that TextBlock has a TextChanged event, I tried with PropertyChanged looking for the Text property in the event args but in the Text property event none of the TextLines had the HasCollapsed bool set to true. Without looking at the Avalonia source I suspect the trimming might be happening at a lower level after this event?
